src/igraph_eulerian.c

```diff
@@ -0,0 +1,614 @@
+                *has_path = 0;
+                *has_cycle = 0;
+                igraph_vector_destroy(&in_degree);
+                igraph_vector_destroy(&out_degree);
+                IGRAPH_FINALLY_CLEAN(2);
+
+                return IGRAPH_SUCCESS;
+            }
+        }
+    }
+
+    IGRAPH_CHECK(igraph_vector_init(&csize_weak, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &csize_weak);
+
+    IGRAPH_CHECK(igraph_clusters(graph, NULL, &csize_weak, &clusters_weak, IGRAPH_WEAK));
+    cluster_count = 0;
+    for (i = 0; i < igraph_vector_size(&csize_weak); i++) {
+        if (VECTOR(csize_weak)[i] > 1) cluster_count++;
+    }
+
+    if (cluster_count > 1) {
+        res_weak = 0;
+    }
+
+    IGRAPH_CHECK(igraph_vector_init(&csize_strong, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &csize_strong);
+
+    IGRAPH_CHECK(igraph_clusters(graph, NULL, &csize_strong, &clusters_strong, IGRAPH_STRONG));
+
+    /* if the number of clusters is the same for strong and weak, and the 
+    graph is weakly connected, then this graph is strongly connected */
+    if (!res_weak || (clusters_strong != clusters_weak)) {
+        res_strong = 0;
+    }
+
+    igraph_vector_destroy(&csize_strong);
+    igraph_vector_destroy(&csize_weak);
+    igraph_vector_destroy(&in_degree);
+    igraph_vector_destroy(&out_degree);
+    IGRAPH_FINALLY_CLEAN(4);
+
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_strong)) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    } else if ((outgoing_excess == 1 && incoming_excess == 1) && 
+        (res_strong || res_weak)) {
+        *has_path = 1;
+        *has_cycle = 0;
+        return IGRAPH_SUCCESS;
+    }
+
+    *has_path = 0;
+    *has_cycle = 0;
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/**
+ * \ingroup Eulerian
+ * \function igraph_is_eulerian
+ * \brief Checks whether an Eulerian path or cycle exists
+ *
+ * An Eulerian path traverses each edge of the graph precisely once. A closed
+ * Eulerian path is referred to as an Eulerian cycle.
+ *
+ * \param graph The graph object.
+ * \param has_path Pointer to a Boolean, will be set to true if an Eulerian path exists.
+ * \param has_cycle Pointer to a Boolean, will be set to true if an Eulerian cycle exists.
+ * \return Error code:
+ *         \c IGRAPH_ENOMEM, not enough memory for
+ *         temporary data.
+ *
+ * Time complexity: O(|V|+|E|), the
+ * number of vertices times the number of edges.
+ *
+ */
+
+int igraph_is_eulerian(igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+    if (igraph_is_directed(graph)) {
+        IGRAPH_CHECK(igraph_i_is_eulerian_directed(graph, has_path, has_cycle));
+    } else {
+        IGRAPH_CHECK(igraph_i_is_eulerian_undirected(graph, has_path, has_cycle));
+    } 
+    return IGRAPH_SUCCESS;
+}
+
+
+int igraph_i_eulerian_path_undirected_implementation(igraph_integer_t start_node, igraph_t *graph, igraph_vector_t *res, igraph_vector_t *outgoing_list) {
```

Comment:
Let's always make the "this" argument the first one. That's the `igraph_t *graph` in this case.